### PR TITLE
Implement instanceID method

### LINKED ISSUE #74: ProviderID should be provided to successfully initialise nodes
Hi,

I tried to use this CCM in a test cluster and I observed that the nodes did not get initialised i.e. taint `node.cloudprovider.kubernetes.io/uninitialized` was never removed. This is the root issue:

```
E0629 09:47:05.321556       1 node_controller.go:364] failed to set node provider id: failed to get instance ID from cloud provider: unimplemented
```

In fact checking the code I noticed that method [InstanceID](https://github.com/kubernetes-sigs/cloud-provider-huaweicloud/blob/f16d76d/pkg/cloudprovider/huaweicloud/instances.go#L103) is not implemented (among others). After manually patching the node to add the `ProviderID` it was initialised successfully.

I think that this is necessary to be able to use this CCM as either manually patching the nodes or setting explicitly the kubelet flag `--provider-id` don't seem to be viable options.

Additional info:
Kubernetes version: `v1.17.5`

### PR DESCRIPTION
**Pull request by irozzo-1A**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR implements the cloud provider `InstanceID` method that is called by the [node controller](https://github.com/kubernetes/kubernetes/blob/3a51aaf8b4baff1242ea30b08d599d913fe7a127/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L522). This allows the users of this CCM not to specify the `provider-id` kubelet flag.

**Which issue(s) this PR fixes**:

Fixes #74 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/cloudprovider/huaweicloud/instances_test.go [hunk 1]
```diff
@@ -68,3 +77,90 @@ func TestParseInstanceTypeFromServerInfo(t *testing.T) {
...
+type ServerResponse struct {
+	Servers []servers.Server `json:"servers"`
+}
+
+func MockListServers(t *testing.T, servers []servers.Server) {
+	res, err := json.Marshal(ServerResponse{Servers: servers})
+	if err != nil {
+		t.Fatalf("Error occurred while marshaling the response: %v", err)
+	}
+	// Handle server creation requests.
```
Comment (irozzo-1A) [inline pkg/cloudprovider/huaweicloud/instances_test.go:151]: change to `Handle server list requests.`

File: pkg/cloudprovider/huaweicloud/instances.go [hunk 2]
```diff
@@ -264,3 +228,93 @@ func (i *Instances) parseInstanceTypeFromServerInfo(srv *servers.Server) (string
...
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(serverList) == 0 {
+		return nil, ErrNotFound
+	}
+
+	return &serverList[0], nil
```
Comment (Dingshujie) [inline pkg/cloudprovider/huaweicloud/instances.go:280]: servers can have same hostname, can't guarantee serverList[0]  is correct one

**Comment by RainbowMango** [comment]
@irozzo-1A 
Thanks for your PR. Have you tested it locally?

**Comment by irozzo-1A** [comment]
> @irozzo-1A
> Thanks for your PR. Have you tested it locally?

@RainbowMango I tested it on a cluster I created, but not locally following [this](https://github.com/kubernetes-sigs/cloud-provider-huaweicloud/blob/master/docs/quick-start-with-local-cluster.md) doc.  Node initialisation went through smoothly. I'm still struggling with an authorization issue when trying to create a service of type LoadBalancer, I don't know yet if it's a configuration issue or a bug, but in case I will file another issue. I don't see any evident relationship with my change though. 

Regarding this PR there is something that I would like to check with you. [Here](https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#node-name-4) is mentioned that `--hostname-override` should be set to the node private IP. What is the reason behind? I think that this does not play nice with this feature, because the node name should correspond to the actual server name for it to work.

**Comment by irozzo-1A** [comment]
@RainbowMango Small update. I tested it using `hack/local-up-cluster.sh` following the quickstart doc and I confirm that the node get initialized successfully, so no regressions I would say. The only caveat as I anticipated is that if the `--provider-id` flag is not specified the node name should match the server name, so `--hostname-override` should not be specified or being equal to the server name.

**Comment by RainbowMango** [comment]
I just quickly tested it using `hack/local-up-cluster.sh` and set the environment as per [the doc](https://github.com/kubernetes-sigs/cloud-provider-huaweicloud/blob/master/docs/quick-start-with-local-cluster.md#set-environment), except that I removed two variables:
- #export HOSTNAME_OVERRIDE=xxx
- #export KUBELET_PROVIDER_ID=xxx

I just found the node will be removed by node lifecycle controller:
```
E0702 11:04:10.183963    5918 node_controller.go:428] failed to get instance ID from cloud provider: instance not found
E0702 11:04:11.399815    5918 node_controller.go:428] failed to get instance ID from cloud provider: instance not found
I0702 11:04:11.930116    5918 instances.go:175] InstanceShutdownByProviderID is called. input provider ID: 
E0702 11:04:12.095204    5918 node_controller.go:428] failed to get instance ID from cloud provider: instance not found
E0702 11:04:12.656009    5918 instances.go:179] Get server info failed. provider id: , error: error occurred while getting server with server id: , error: {"ErrorCode":"APIGW.0101","Message":"The API does not exist or has not been published in the environ}
E0702 11:04:12.656035    5918 node_lifecycle_controller.go:154] error checking if node 192.168.1.122 is shutdown: error occurred while getting server with server id: , error: {"ErrorCode":"APIGW.0101","Message":"The API does not exist or has not been publ}
I0702 11:04:13.077677    5918 node_lifecycle_controller.go:180] deleting node since it is no longer present in cloud provider: 192.168.1.122
```

Kubernetes version:
```
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.0-beta.2.577+180af4240c4290-dirty", GitCommit:"180af4240c42906f072112e7ac7818f4b3574c5a", GitTreeState:"dirty", BuildDate:"2020-07-02T03:03:15Z", GoVersion:"go1.14.4", Compiler:"gc", P}
```

/cc @Dingshujie 
Involve @Dingshujie here, she is an expert with a lot of production experience.

File: pkg/cloudprovider/huaweicloud/instances.go [hunk 3]
```diff
@@ -100,9 +110,17 @@ func (i *Instances) NodeAddressesByProviderID(ctx context.Context, providerID st
...
 // cloudprovider.InstanceNotFound should NOT be returned for instances that exist but are stopped/sleeping
+//
+// Based on: https://github.com/kubernetes/cloud-provider-openstack/blob/2493d93/pkg/cloudprovider/providers/openstack/openstack_instances.go#L162
 func (i *Instances) InstanceID(ctx context.Context, nodeName types.NodeName) (string, error) {
-	klog.Infof("InstanceID is called. input nodeName: %s", nodeName)
-	return "", cloudprovider.NotImplemented
+	srv, err := i.getServerByName(nodeName)
+	if err != nil {
+		if err == ErrNotFound {
+			return "", cloudprovider.InstanceNotFound
```
Comment (RainbowMango) [inline pkg/cloudprovider/huaweicloud/instances.go:119]: I'd prefer adding one error log when returns an error.

Same as below.

**Comment by irozzo-1A** [comment]
I just quickly tested it using hack/local-up-cluster.sh and set the environment as per the doc, except that I removed two variables:

#export HOSTNAME_OVERRIDE=xxx
#export KUBELET_PROVIDER_ID=xxx

@RainbowMango sorry if I did not mention it, but this is not enough to test it with `hack/local-up-cluster.sh` because if you don't specify those env variables `--hostname-override` [defaults to `127.0.0.1`](https://github.com/kubernetes/kubernetes/blob/908847c/hack/local-up-cluster.sh#L77) and `--provider-id` [defaults to the hostname](https://github.com/kubernetes/kubernetes/blob/908847c/hack/local-up-cluster.sh#L83). If you want to test it that way you should do the following:

- Removed those lines: https://github.com/kubernetes/kubernetes/blob/908847c/hack/local-up-cluster.sh#L712,L713
- `export HOSTNAME_OVERRIDE="$(hostname)"`

File: pkg/cloudprovider/huaweicloud/instances.go (hunk 3 above)
Comment (irozzo-1A) [reply pkg/cloudprovider/huaweicloud/instances.go:119]: note that it gets eventually logged by the node controller, e.g.

```
E0702 07:24:11.537080   20857 node_controller.go:428] failed to get instance ID from cloud provider: instance not found
```

Personally I would find it a bit redundant in this case, but it does not cost me much effort to add it if you still prefer ;-)

**Comment by RainbowMango** [comment]
Yeah, you are right.
/lgtm
/approve

Thanks again for your contribution :). Just feel free to send a PR if you want to improve it.

File: pkg/cloudprovider/huaweicloud/instances.go (hunk 3 above)
Comment (RainbowMango) [reply pkg/cloudprovider/huaweicloud/instances.go:119]: It's no big deal. No worries.

**Pull request merged**